### ISSUE TITLE
Python3 version of NVDA fails to start on clean install of Windows 7 x64

### ISSUE THREAD

**Issue by lukaszgo1** [nvaccess/nvda#10031]
### Steps to reproduce:
1. Perform clean install of Windows 7 x64 sp1.
2. Try to start NVDA version using Python 3 either from the launcher, or from the potable copy.
### Actual behavior:
When using launcher starting tone plays for a while and simply stops playing nothing else happens. When using portable version dialog box with the following error translated from Polish appears:

```
---------------------------
nvda.exe - System error
---------------------------
Cannot start the program because  api-ms-win-crt-runtime-l1-1-0.dll is not found on your computer. Try reinstalling the program to fix this issue.
---------------------------
OK   
---------------------------
```

### Expected behavior:
NVDA should start
#### NVDA installed/portable/running from source:
Tried from launcher and portable.
#### NVDA version:
alpha-18290,7b5cd2d0
#### Windows version:
Clean install of Windows 7 pro x64
#### Name and version of other software in use when reproducing the issue:
None
#### Other information about your system:
On my two other machines running Windows 7 installed from the samae iso image this isn't happening, so it is the clean environment which is causing the problem.
#### Does the issue still occur after restarting your PC?
Yes
#### Have you tried any other versions of NVDA? If so, please report their behaviors.
yes:
- With NVDA version alpha-18064,e579eb02 that is last Python 2 version it works on the same machine.
- with NVDA version alpha-18178,ba6b4f6b that is the first Python 3 version I can reproduce this.

**Comment by lukaszgo1** [comment]
cc @feerrenrut @leonardder @josephsl I believe this is a good candidate for a  P1.

**Comment by leonardder** [comment]
Is this windows 7 sp1?

**Comment by lukaszgo1** [comment]
Yes, it is.

**Comment by leonardder** [comment]
Could you please install python 3.7.4 on this system to see wheter it is able to start?

**Comment by lukaszgo1** [comment]
Some more info:
1. This is caused by the fact that apps written in Python 3.6 and later require Universal  CRT to be present. Installing Python itself installs required files - I've tried on the affected machine and Python installation allowed NVDA to start correctly.
2. Universal CRT is present by default on Windows 10 (all builds according to [this blog post](https://devblogs.microsoft.com/cppblog/introducing-the-universal-crt/)). On earlier versions of Windows it is distributed via Windows Update, hence the failure on a clean install.

I see two possibilities of fixing this:
1. State in the user guide that installation of KB2999226 is required on versions of Windows older than Windows 10.
2. Distribute needed dll's ourselves. They are present in C:\Program Files (x86)\Windows Kits\10\Redist\ucrt\DLLs\x86 if Windows 10 SDK is installed. The blog post linked above recommends to copy all of them. The only downside is the fact that NVDA directory would be quite messy after that - there are 42 of them. I've done it manually, and can confirm that it allows NVDA to start on the machine without Universal CRT installed. Doing it as part of scons seems doable.


I believe we want to support NVDA in as many situations as possible including clean installs of Windows. This affects not only Windows 7 but also 8.1 and server versions of these systems, and we cannot assume that all users have the needed update installed.

**Comment by leonardder** [comment]
Py2exe actually bundles these files by default, but we manually exclude them. This seems pretty intentional.May be @jcsteh knows more?

**Comment by lukaszgo1** [comment]
After reading the [commit](https://github.com/nvaccess/nvda/commit/dc6e39071fbc253416f6c5017f1cc334f7fdf8c1) in which excluding was introduced I believe that the intention was not to bundle a dll from whatever version of Windows was in use at the time of build which is more than understandable, If we would exclude dlls from the system and as part of scons source copy the libraries from Windows 10 SDk this should work. In addition the commit talks about problems with starting NVDA on Windows XP, so the situation might have changed dramatically  since then.

**Comment by jcsteh** [comment]
The original reason for this is that certain system dlls from later versions of Windows (e.g. powrprof.dll) depend on other dlls that don't exist on earlier versions of Windows. Because the appropriate version of powrprof.dll (for example) is bundled with Windows, there's no need to include it, and including it ends up breaking things.

I think this was before the Universal CRT was... universal. So, yes, the situation has changed. We could either copy in the dlls from the SDK or just remove the blacklisting of api-* dlls in setup.py. The latter would ensure we don't bundle files we don't need, but would need to be tested on a Windows 7 system with a build from Windows 10 just to be sure. We should not remove blacklisting for any other dlls, though.

Note that this will increase the size of the NVDA download somewhat, though hopefully not too much.

**Comment by leonardder** [comment]
@lukaszgo1: Here is a try build that should fix this: https://ci.appveyor.com/api/buildjobs/v99dyh533g9wl62g/artifacts/output%2Fnvda_snapshot_try-dontStripUniversalCRT-18301%2C233f9e42.exe

**Comment by Brian1Gaff** [comment]
For the record here, I had no issues and I see 31 versions of this file and 
its support files in various folders. Of course as has been said, it could 
well be that either windows update has installed them or another piece of 
software using Python 3. This is kind of the same issue  found with the.NET 
framework, often such software spots  that this is missing and installs it 
from Microsoft web site, could this be done here?
 As a Matter of interest, if you leave Windows does it download it 
eventually?
 One possible issue is how forcefully  Microsoft enforce end of life for 
Windows 7 next year of course. This could break any solution  based on 
downloading the KB update.
 Brian

bglists@blueyonder.co.uk
Sent via blueyonder.
Please address personal E-mail to:-
briang1@blueyonder.co.uk, putting 'Brian Gaff'
in the display name field.
Newsgroup monitored: alt.comp.blind-users
-----

**Comment by lukaszgo1** [comment]
> 
> For the record here, I had no issues and I see 31 versions of this file and its support files in various folders. Of course as has been said, it could well be that either windows update has installed them or another piece of software using Python 3. This is kind of the same issue found with the.NET framework, often such software spots that this is missing and installs it from Microsoft web site, could this be done here?

I don't see how it could be done. How are you going to install it without a screen reader running?


>  As a Matter of interest, if you leave Windows does it download it eventually?

Yes, it is a recommended update.

**Comment by leonardder** [comment]
I think bundling all these files from the Windows 10 sdk is best what we can do. SCons currently finds them in `C:\Program Files (x86)\Microsoft Visual Studio\2017\Community\Common7\IDE`

**Comment by lukaszgo1** [comment]
@leonardder The build that you linked to still fails with the following error:

The entry point of function ucrtbase.abort cannot be found in the library api-ms-win-crt-runtime-l1-1-0.dll.
Bundling all Universal CRT files seems easier indeed.

**Comment by leonardder** [comment]
@michaeldcurran: would you be the right person to look into this?

**Comment by leonardder** [comment]
Probably good to combine this with #10238 when addressing this. In short, we probably want to ensure that scons copies brlapi as well, as py2exe might think that brlapi is a system dll when brltty is installed.
